src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -299,6 +314,87 @@ unsigned FindNextDeviceTypeIndex(Span<const EmberAfDeviceType> types, const Data
 
 const ConcreteCommandPath kInvalidCommandPath(kInvalidEndpointId, kInvalidClusterId, kInvalidCommandId);
 
+std::optional<DataModel::EndpointInfo> GetEndpointInfoAtIndex(uint16_t endpointIndex)
+{
+    VerifyOrReturnValue(emberAfEndpointIndexIsEnabled(endpointIndex), std::nullopt);
+    EndpointId parent = emberAfParentEndpointFromIndex(endpointIndex);
+    if (GetCompositionForEndpointIndex(endpointIndex) == EndpointComposition::kFullFamily)
+    {
+        return DataModel::EndpointInfo(parent, DataModel::EndpointCompositionPattern::kFullFamilyPattern);
+    }
+    if (GetCompositionForEndpointIndex(endpointIndex) == EndpointComposition::kTree)
+    {
+        return DataModel::EndpointInfo(parent, DataModel::EndpointCompositionPattern::kTreePattern);
+    }
+    return std::nullopt;
+}
+
+DataModel::EndpointEntry FirstEndpointEntry(unsigned start_index, uint16_t & found_index)
+{
+    // find the first enabled index after the start index
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+    for (uint16_t endpoint_idx = static_cast<uint16_t>(start_index); endpoint_idx < lastEndpointIndex; endpoint_idx++)
+    {
+        if (emberAfEndpointIndexIsEnabled(endpoint_idx))
+        {
+            found_index                            = endpoint_idx;
+            DataModel::EndpointEntry endpointEntry = DataModel::EndpointEntry::kInvalid;
+            endpointEntry.id                       = emberAfEndpointFromIndex(endpoint_idx);
+            auto endpointInfo                      = GetEndpointInfoAtIndex(endpoint_idx);
+            // The endpoint info should have value as this endpoint should be valid at this time
+            VerifyOrDie(endpointInfo.has_value());
+            endpointEntry.info = endpointInfo.value();
+            return endpointEntry;
+        }
+    }
+
+    // No enabled endpoint found. Give up
+    return DataModel::EndpointEntry::kInvalid;
+}
+
+bool SemanticTagsEqual(const DataModel::Provider::SemanticTag & tagA, const DataModel::Provider::SemanticTag & tagB)
+{
+    if (tagA.label.HasValue() != tagA.label.HasValue())
+    {
+        return false;
+    }
+    if (tagA.label.HasValue())
+    {
+        if (tagA.label.Value().Value().size() != tagB.label.Value().Value().size())
+        {
+            return false;
+        }
+        if (strncmp(tagA.label.Value().Value().data(), tagB.label.Value().Value().data(), tagB.label.Value().Value().size()) != 0)
+        {
+            return false;
+        }
+    }
+    return (tagA.tag == tagB.tag) && (tagA.mfgCode == tagB.mfgCode) && (tagA.namespaceID == tagB.namespaceID);
+}
+
+std::optional<unsigned> FindNextSemanticTagIndex(EndpointId endpoint, const DataModel::Provider::SemanticTag & previous,
```

Comment:
please add some doccomment about this. It seems that this returns nullopt if definately not found, but may also return an out of range index since it returns `+1` whenever it finds the previous match. 

Could we add a check so this never returns out of range?

Reply:
Added a check here.